src/transformStateWithClones.js

```diff
@@ -6,8 +6,30 @@
  *
  * @return {Object[]}
  */
+
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  actions.forEach((action) => {
+    if (action.type === 'clear') {
+      currentState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(currentState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties' && action.keysToRemove) {
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    }
+    stateHistory.push({ ...currentState });
```

Comment:
Consider using a `switch` statement instead of multiple `if` statements when you have a limited amount of conditions to check. This will improve code readability and maintainability. Remember to add a `default` case for error handling as well.